add support for map type

ananthakumaran opened the pull request:
I am not sure about the public API exposed by the library, so tried to keep all the exported function backward compatible.

k32 commented:
Thanks for your PR! Could you please bump app version and add a changelog entry?

zmstone commented on src/avro_binary_decoder.erl:
```diff
@@ -134,27 +149,33 @@ dec(Bin, T, _Lkup, Hook) when ?IS_FIXED_TYPE(T) ->
       end, {[], Bin}, FieldTypes),
-  FieldValues = lists:reverse(FieldValuesReversed),
-  {FieldValues, Tail}.
+  FieldValues0 = lists:reverse(FieldValuesReversed),
```
no need to reverse for `maps:from_list/1`. i.e. put this to the `proplist` clause.

zmstone commented on src/avro.erl:
```diff
@@ -130,6 +133,7 @@
              | [{name(), avro:out()}].
 
 -type codec_options() :: [proplists:property()].
+-type decoder_options() :: map().
```
there is no need to add another type, please update `codec_options()` instead

ananthakumaran replied:
Could you clarify this comment a little more? Do you suggest to use proplist instead of map for decoder options or you meant adding something like ` | map()` in the dialzyer spec.

zmstone replied:
ok. now I realise that it's perhaps not a good idea to have encoder/decoder options mixed.
Please update `codec_options()` to `encoder_options() :: prolists:proplist()`
And keep `decoder_options()` as it is, however, public APIs should still allow proplist as input,
so we need to either update API specs with `proplists:proplist() | decoder_options()`
or make another union type for it.

Please also be more specific about map fields.
i.e. write this type spec as:
```
-type decoder_options() :: #{ encoding => avro_binary | avro_json,
                            , map_type => proplist | map
                            , ....
                            }
```

zmstone replied:
maybe export `-type codec_options() :: encoder_options() | decoder_options().` to be backward compatible
but we do not use it in our specs anymore.

zmstone replied:
request for changes @ananthakumaran

ananthakumaran replied:
I have moved the definition to avro_internal.hrl and did some changes so we there won't be any need to expose the structure to the user. The user will still supply the option in the old code_options() type and that will be cast to maps and passed to other modules.

ananthakumaran commented on src/avro_json_decoder.erl:
```diff
@@ -119,15 +121,18 @@ decode_schema(JSON, Opts) when is_list(Opts) ->
-  parse(DecodedJson, Schema, Lkup, IsWrapped, Hook).
+  %% parse to wrapped value by default
+  IsWrapped = maps:get(is_wrapped, Options, true),
+  parse(DecodedJson, Schema, Lkup, IsWrapped, Hook, Options).
```
is there any rule of thumb I can follow to figure out what is exactly considered as Public API? are all exported functions considered as public API?

I was mostly trying to keep the backward compatiblity without changing the structure much. That's what probably makes the logic weird in some places. If you check the call chain, Hook is already present in the Options, so strictly speaking, we don't to need to pass it downstream as a separate arg either. Should I go all out and remove Hook and IsWrapped and add a fallback for old api, some like below

```erlang
parse(Value, TypeName, Lkup, IsWrapped, Hook) ->
  parse(Value, TypeName, Lkup, avro:make_decoder_options([{hook, Hook}, {is_wrapped, Wrapped}])).

parse(Value, TypeName, Lkup, Options) ->
```

zmstone replied:
Only APIs exported from avro.erl are considered public, feel free to change internal APIs.
Agree to put both `IsWrapped` and `Hook` in options.

zmstone commented:
unfortunately, we have a limit of maximum 80-char line width.
please help to make the long lines shorter

zmstone commented on src/avro_record.erl:
```diff
@@ -339,6 +339,8 @@ to_term(Record) when ?IS_RECORD_VALUE(Record) ->
 -spec encode(record_type(), [{field_name_raw(), avro:in()}],
              fun((field_name(), avro_type(), avro:in()) -> avro:out())) ->
         [avro:out()].
+encode(Record, Map, EncodeFun) when is_map(Map) ->
```
and here too

ananthakumaran replied:
I have done some changes in both Record and Map. It's still not as clean as I would like. I can't find any api to map over both key and value, `maps:map` only maps the value, but we have to canonicalize the key, so I had to cast to list and then cast back to map.

I tried to use map as the internal value for record, but the arbitrary order of maps causes issues. Other modules seem to assume that `avro_record:to_list` will return the values in correct field order

ananthakumaran commented:
I thought `make elvis-rock` would warn me about line length, I did fix all the errors reported by it. Btw, I didn't mean to trigger another review as I still need to address some of the review comments. I will ping you once I addressed all the comments.

zmstone commented:
elvis does not check test modules.

zmstone commented:
this diff should make elvis check `test` dir, please help to add it:
```diff --git a/elvis.config b/elvis.config
index c1391d2..7a420db 100644
--- a/elvis.config
+++ b/elvis.config
@@ -57,7 +57,15 @@
                    , {elvis_style, no_debug_call}

                  ]
-       }
+       },
+      #{dirs => ["test"],
+        filter => "*.erl",
+        rules => [ {elvis_style, line_length,
+                      #{ limit => 80,
+                         skip_comments => false
+                       }}
+                 ]
+      }
      ]
     }
    ]```

zmstone commented:
@ananthakumaran Perfect PR!
